custom/ilsgateway/slab/views.py

```diff
@@ -0,0 +1,115 @@
+
+from corehq.apps.domain.decorators import domain_admin_required
+from corehq.apps.domain.views import BaseDomainView
+from corehq.apps.locations.models import Location, SQLLocation
+from corehq.apps.locations.views import LocationsListView
+from corehq.apps.reports.standard import CustomProjectReport
+from corehq.apps.style.decorators import use_jquery_ui, use_bootstrap3
+from custom.ilsgateway.models import SLABConfig
+from custom.ilsgateway.slab.forms import SLABEditLocationForm
+from dimagi.utils.decorators.memoized import memoized
+
+
+class SLABConfigurationView(LocationsListView):
+    template_name = 'ilsgateway/slab/slab_configuration.html'
+
+    @use_bootstrap3
+    @use_jquery_ui
+    @method_decorator(domain_admin_required)
+    def dispatch(self, request, *args, **kwargs):
+        return super(SLABConfigurationView, self).dispatch(request, *args, **kwargs)
+
+
+class SLABEditLocationView(BaseDomainView):
+    template_name = 'ilsgateway/slab/edit_location.html'
+    page_title = 'Edit location'
+
+    @use_bootstrap3
+    @use_jquery_ui
+    @method_decorator(domain_admin_required)
+    def dispatch(self, request, *args, **kwargs):
+        if self.sql_location.location_type.administrative:
+            return HttpResponseBadRequest()
+        return super(SLABEditLocationView, self).dispatch(request, *args, **kwargs)
+
+    @property
+    @memoized
+    def location(self):
+        return Location.get(self.kwargs['location_id'])
+
+    @property
+    @memoized
+    def sql_location(self):
+        return self.location.sql_location
+
+    def section_url(self):
+        return reverse(
+            'slab_edit_location',
+            kwargs={'domain': self.domain, 'location_id': self.kwargs['location_id']}
+        )
+
+    @property
+    def page_context(self):
+        try:
+            slab_config = SLABConfig.objects.get(sql_location=self.sql_location)
+            is_pilot = slab_config.is_pilot
+            selected_ids = slab_config.closest_supply_points.all().values_list('pk', flat=True)
+        except SLABConfig.DoesNotExist:
+            is_pilot = False
+            selected_ids = []
+        form = SLABEditLocationForm(initial={'is_pilot': is_pilot, 'selected_ids': selected_ids})
+        form.fields['selected_ids'].choices = self.choices
+        return {
+            'form': form
+        }
+
+    @property
+    def choices(self):
+        return SLABConfig.objects.filter(is_pilot=True) \
+            .exclude(sql_location=self.sql_location).values_list('sql_location__pk', 'sql_location__name')
+
+    def post(self, request, *args, **kwargs):
+        form = SLABEditLocationForm(data=request.POST)
+        form.fields['selected_ids'].choices = self.choices
+        if not form.is_valid():
+            return self.get(request, *args, **kwargs)
+
+        cleaned_data = form.clean()
+        is_pilot = cleaned_data['is_pilot']
+        selected_ids = cleaned_data['selected_ids']
+        if is_pilot:
```

Comment:
we should put everything in the is_pilot block in a CriticalSection. we want to control access to this critical section by location id right? if so then it would just be:
`with CriticalSection(['update-slab-config-for-location-%s' % self.sql_location.location_id]):`